ui/app/hooks/useTransactionDisplayData.js

```diff
@@ -0,0 +1,118 @@
+const INTERACTION = 'interaction'
+const APPROVAL = 'approval'
+
+// This is duplicated from transactions selectors, additionally
+// from the history view we will already know the pending status
+// of a transaction, but in order to make this hook less coupled
+// to where it is used we should determine the pending status here
+const pendingStatusMap = {
+  [APPROVED_STATUS]: true,
+  [SUBMITTED_STATUS]: true,
+}
+
+/**
+ * @typedef {Object} TransactionDisplayData
+ * @property {string} title               - primary description of the transaction
+ * @property {string} subtitle            - supporting text describing the transaction
+ * @property {string} category            - the transaction category
+ * @property {string} primaryCurrency     - the currency string to display in the primary position
+ * @property {string} [secondaryCurrency] - the currency string to display in the secondary position
+ * @property {string} status              - the status of the transaction
+ * @property {string} senderAddress       - the Ethereum address of the sender
+ * @property {string} recipientAddress    - the Ethereum address of the recipient
+ */
+
+/**
+ * useTransactionDisplayData
+ *
+ * The goal of this method is to perform all of the necessary computation and
+ * state access required to take a transactionGroup and derive from it a shape
+ * of data that can power all views related to a transaction. Presently the main
+ * case is for shared logic between transaction-list-item and transaction-detail-view
+ * @param {Object} transactionGroup group of transactions
+ * @return {TransactionDisplayData}
+ */
+export function useTransactionDisplayData (transactionGroup) {
+  const knownTokens = useSelector(tokenSelector)
+  const t = useI18nContext()
+  const { initialTransaction, primaryTransaction } = transactionGroup
+  // initialTransaction contains the data we need to derive the primary purpose of this transaction group
+  const { transactionCategory } = initialTransaction
+
+  const { from: senderAddress, to } = initialTransaction.txParams || {}
+
+  // for smart contract interactions, methodData can be used to derive the name of the action being taken
+  const methodData = useSelector((state) => getKnownMethodData(state, primaryTransaction?.txParams?.data)) || {}
+
+  const actionKey = getTransactionActionKey(initialTransaction)
+  const statusKey = getStatusKey(primaryTransaction)
+  const status = statusKey in pendingStatusMap ? 'pending' : statusKey
+
+  const primaryValue = primaryTransaction.txParams?.value
+  let prefix = '-'
+  const date = formatDateWithYearContext(initialTransaction.time || 0)
+  let subtitle = date
+  let recipientAddress = to
+
+  const token = knownTokens.find((token) => token.address === recipientAddress)
+  const tokenData = useTokenData(initialTransaction?.txParams?.data)
+  const tokenDisplayValue = useTokenDisplayValue(initialTransaction?.txParams?.data, token)
+
+  let category = SEND
+  let title = t('sendETH')
```

Comment:
Rather than leaving this as the default, maybe we should ensure we exhaustively check all transaction categories? It seems dangerous to risk unidentified things being shown as a send.

Reply:
Is there an exhaustive list of transaction categories that could be seen here? I tried to piece them together through the various transforms that happen before we get to this point in the logic flow. I'm sure I'm missing more in retrospect. eg: what does a 'signing request' category look like?

Reply:
I think this one function is the only place that `transactionCategory` is set for outgoing transactions: https://github.com/MetaMask/metamask-extension/blob/develop/app/scripts/controllers/transactions/index.js#L692
The the category is always `incoming` for incoming transactions (handled in the `incoming-transactions` controller).

Just 7 categories. I think those are all accounted for here in these 4 (as _send_ and _receive_ each include both ETH and tokens, and a contract deployment is displayed as a " contract interaction").

Then there are the non-transactions (which have no `transactionCategory`). I think this just includes signature requests, but I'm not 100% certain 🤔

Reply:
It looks like signature requests are indeed the only non-transaction things included in the list. The transaction selector brings them in via `unapprovedMessagesSelector` - those are all signature requests.

Reply:
```js
  let category
  let title
  // There are four types of transaction entries that are currently differentiated in the design
  // 1. Send (sendEth sendTokens)
  // 2. Deposit
  // 3. Site interaction
  // 4. Approval
  if (transactionCategory == null) {
    category = APPROVAL
    title = t('signatureRequest')
    subtitle += ` · ${initialTransaction.origin}`
  } else if (transactionCategory === 'approve') {
    category = APPROVAL
    title = t('approveSpendLimit')
    subtitle += ` · ${initialTransaction.origin}`
  } else if (transactionCategory === 'contractDeployment' || transactionCategory === 'contractInteraction') {
    category = INTERACTION
    title = (methodData?.name && camelCaseToCapitalize(methodData.name)) || (actionKey && t(actionKey)) || ''
    subtitle += ` · ${initialTransaction.origin}`
  } else if (transactionCategory === 'incoming') {
    category = DEPOSIT
    title = t('deposit')
    prefix = ''
    subtitle += ` · From: ${shortenAddress(senderAddress)}`
  } else if (transactionCategory === 'transfer') {
    category = SEND
    title = t('sendSpecifiedTokens', [token?.symbol])
    recipientAddress = getTokenToAddress(tokenData.params)
    subtitle += ` · To: ${shortenAddress(recipientAddress)}`
  } else if (category === SEND) {
    category = SEND
    title = t('sendETH')
    subtitle += ` · To: ${shortenAddress(recipientAddress)}`
  }
```

this is the adjustment I've made. Not sure if signing requests should be approvals or contract interactions for now, or hold off for new direction from @rachelcope on a new symbol to display. 

two questions @Gudahtt 
1: do you think, pending design feedback, this covers our bases.
2: This isn't much better than setting category, subtitle and title to the 'signature request' as defaults... should we find a way to update transactionCategory for signature requests throughout the system so that its not a fallback?

Reply:
Until we get a better icon to use, I'd suggest using the icon we use today for signature requests for now (the Ethereum logo). As inappropriate and non-explanatory as the Ethereum logo is for these, each of these new four would be equally inappropriate, so :man_shrugging: 

Assuming anything without a category is a signature request seems fine to me for now. Certainly assigning them a transaction category seems inappropriate. A better solution might be creating a separate `ListItem` for signature requests, and splitting up the selectors and everything to treat them as separate. But that's a ton of work that we can think about some other day.

RE: Having an exhaustive list of transactions, I mainly wanted to make sure we weren't missing any categories.

Reply:
![image](https://user-images.githubusercontent.com/4448075/82685837-77ba4400-9c1a-11ea-8aca-14d433f0bde2.png)